src/com/zenuml/dsl/PsiToDslConverter.java

```diff
@@ -30,113 +34,75 @@ public void visitMethod(PsiMethod method) {
     private void visitMethod(PsiMethod method, String insertBefore) {
         LOG.debug("Enter: visitMethod: " + method);
 
-        if (callStack.contains(method)) {
-            String callLoop = Stream.concat(callStack.stream(), Stream.of(method))
-                    .map(PsiMethod::getName)
-                    .reduce((m1, m2) -> format("%s -> %s", m1, m2))
-                    .get();
-            LOG.info(format("Call loop detected: %s, stopped", callLoop));
+        if (methodStack.contains(method)) {
+            LOG.debug("Exit (loop detected): visitMethod: " + method);
+            zenDsl.changeLine();
             return;
         }
 
-        String methodName = getMethodPrefix(method) + method.getName();
+        String methodCall = getMethodCall(method);
 
-        String remainder = "";
-        if (insertBefore != null) {
-            int index = dsl.lastIndexOf(insertBefore);
-            remainder = this.dsl.substring(index);
-            this.dsl = this.dsl.substring(0, index);
-        }
-
-        this.dsl += methodName;
-
-        callStack.add(method);
-
-        super.visitMethod(method);
-        callStack.remove(method);
-
-        if (remainder.length() > 0) {
-            this.dsl += getIndent(level - 1) + remainder;
-        }
+        process(method, insertBefore, methodCall);
 
         LOG.debug("Exit: visitMethod: " + method);
     }
 
-    private String getMethodPrefix(PsiMethod method) {
-        int size = callStack.size();
-        if (size > 0) {
-            PsiMethod parentMethod = callStack.get(size - 1);
-            if (parentMethod.getContainingClass().equals(method.getContainingClass())) {
-                return "";
-            }
+    private void process(PsiMethod method, String insertBefore, String methodCall) {
+        if (insertBefore == null) {
+            zenDsl.addMethodCall(methodCall);
+            processChildren(method);
+        } else {
+            int index = zenDsl.getDsl().lastIndexOf(insertBefore);
```

Comment:
Exposed internals of `zenDsl`, can extract a method like `zenDsl.insertBefore((zenDsl) -> processChildren(method), insertBefore)`.